### PR TITLE
Virtualize the S4TF runtime interface. 

### PR DESCRIPTION
**Pull request by bgogul**
An attempt to virtualize the runtime interface. This PR added two protocols `TensorOperation` and `TFTensorOperation`. As the name implies, `TensorOperation` is TensorFlow agnostic, where as `TFTensorOperation` is tensorflow-specific.

This PR also updates the `EagerExecution.swift`,  `RawOpsGenerated.swift`, the `generate_wrappers.py` script and the gyb file to reflect all the necessary changes. 

This partially fixes https://bugs.swift.org/browse/TF-502

### CODE REVIEW

File: TensorFlowGraphOperation.swift [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+/// A handle that is compatible with the TensorFlow library.
+public protocol TensorFlowHandle {
+  var _cTensorHandle: CTensorHandle {get}
+}
+
+extension _AnyTensorHandle : TensorFlowHandle {}
+
+/// A graph operation that is compatible with the Tensor library.
```
Comment (burmako) [inline TensorFlowGraphOperation.swift]:
```suggestion
/// A graph operation that is compatible with the TensorFlow library.
```

File: TensorFlowGraphOperation.swift [hunk 2]
```diff
@@ -0,0 +1,168 @@
+extension _ExecutionContext {
+  // The execution mode is effectively encoded in the GraphOperation.
+  // We can use this to switch switch between different execution modes.
```
Comment (burmako) [inline TensorFlowGraphOperation.swift]:
```suggestion
  // We can use this to switch between different execution modes.
```

File: GraphOperation.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
+public protocol GraphOperation {
```
Comment (burmako) [inline GraphOperation.swift]: Where do you see this protocol ultimately living in the future? It doesn't depend on TensorFlow in any way, right?

File: GraphOperation.swift [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  associatedtype TensorValueHandle
+  // We use functions instead of fields to give freedom in the
+  // representation for the conforming types.
+  init(_ name: String, _ nOutputs: Int)
+
+  // TODO: addInput does not need to return a value, but leaving it
+  // for now so that it is compatible with the generator script.
```
Comment (burmako) [inline GraphOperation.swift]: What is the return value used for currently?

File: GraphOperation.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
+public protocol GraphOperation {
+  associatedtype TensorValueHandle
```
Comment (burmako) [inline GraphOperation.swift]: We cannot use `TensorHandle` because that name is already taken by the stdlib?

File: EagerExecution.swift [hunk 6]
```diff
@@ -21,19 +21,23 @@ import CTensorFlow
...
 
   @usableFromInline
-  internal init(_ name: String) {
+  internal init(_ name: String, _ nOutputs: Int) {
     self.status = TF_NewStatus()
     self.op = TFE_NewOp(_ExecutionContext.global.eagerContext, name, status)
+    self.nOutputs = nOutputs
   }
 
   @inlinable @inline(__always)
```
Comment (burmako) [inline EagerExecution.swift:37]: I noticed that we're often using `@inlinable @inline(__always)`. Is this just an artifact of GPE, or there's another reason to annotate definitions this way?

File: TensorFlowGraphOperation.swift [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    _ count3: Int,
+    _ count4: Int,
+    _ count5: Int,
+    _ count6: Int,
+    _ count7: Int,
+    _ count8: Int,
+    _ count9: Int
+  ) -> (T0, T1, T2, T3, T4, T5, T6, T7, T8, T9)
+
+  // The following should work as well.
```
Comment (burmako) [inline TensorFlowGraphOperation.swift]: How would these functions decide which instances of `TensorArrayProtocol` to create as their return values?

File: RawOpsGenerated.swift [hunk 8]
```diff
@@ -35551,20 +36896,690 @@ public static func xdivy<T: FloatingPoint & TensorFlowScalar>(
...
+) -> T {
+  let nOutputs = Int(input._typeList.count)
+  let op = makeOp("XlaWhile", nOutputs)
+  op.setAttr("T", input._typeList)
+  op.setAttr("cond", cond)
+  op.setAttr("body", body)
+  let _ = op.addInputList(input)
+  return op.execute(Int(input._typeList.count))
+}
+
```
Comment (burmako) [inline RawOpsGenerated.swift]: Where does this big addition come from? Changes to `generate_wrappers.py` don't look related, do they?

File: TensorFlowGraphOperation.swift (hunk 7 above)
Comment (saeta) [reply TensorFlowGraphOperation.swift]: I assume this is using return-type type inference in the generated ops case?

File: GraphOperation.swift (hunk 3 above)
Comment (saeta) [inline GraphOperation.swift]: Hmm, random thought: should we call this something like `TensorOperation` or `TFOpProtocol`? These operations (IIUC) are not necessarily graph operations (e.g. could be eager).

**Review by saeta — APPROVED** [review]

File: EagerExecution.swift (hunk 6 above)
Comment (lattner) [reply EagerExecution.swift:37]: I'm pretty sure this was just for GPE and can be removed.  @rxwei would know for sure.

Comment (rxwei) [reply EagerExecution.swift:37]: Yes, all the `@inlinable`, `@usableFromInline`, `@inline(__always)` can be removed.

File: GraphOperation.swift (hunk 3 above)
Comment (bgogul) [reply GraphOperation.swift]: I wanted this to be not specific for TF. I don't prefer `TFOpProtocol`, but would be OK with changing it to `TensorOperation`

File: GraphOperation.swift (hunk 5 above)
Comment (bgogul) [reply GraphOperation.swift]: Yes, it is taken by `TensorHandle`. I am planning to get rid of `TensorHandle` as well. I can rename `TensorValueHandle` to `TensorHandle` when I get rid of the other one.

File: GraphOperation.swift (hunk 4 above)
Comment (bgogul) [reply GraphOperation.swift]: It is not being used at all. All use site have `let _ = addInput...`. The return values can be removed as far as I can see.

File: TensorFlowGraphOperation.swift (hunk 7 above)
Comment (bgogul) [reply TensorFlowGraphOperation.swift]: As @saeta says.

File: RawOpsGenerated.swift (hunk 8 above)
Comment (bgogul) [reply RawOpsGenerated.swift]: I used a newer version of TensorFlow python library, which presumably has more operations.

**Comment by bgogul** [comment]
I fixed the comments so far and update the related scripts. I am going to merge this in to make further progress. I will be happy to address any comments later.
